packages/tpetra/core/src/Tpetra_Details_extractBlockDiagonal.hpp

```diff
@@ -0,0 +1,106 @@
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+#ifndef TPETRA_DETAILS_EXTRACTBLOCKDIAGONAL_HPP
+#define TPETRA_DETAILS_EXTRACTBLOCKDIAGONAL_HPP
+
+#include "TpetraCore_config.h"
+#include "Tpetra_CrsMatrix.hpp"
+#include "Teuchos_RCP.hpp"
+
+/// \file Tpetra_Details_extractBlockDiagonal.hpp
+/// \brief Functions that allow for the extraction of a block diagonal
+/// from a Tpetra::CrsMatrix.
+/// \warning This file, and its contents, are implementation details
+///   of Tpetra.  The file itself or its contents may disappear or
+///   change at any time.
+
+namespace Tpetra {
+namespace Details {
+
+
+template<class SparseMatrixType,
+         class MultiVectorType>
+void extractBlockDiagonal(const SparseMatrixType& A, MultiVectorType & diagonal) {
+  using local_map_type = typename SparseMatrixType::map_type::local_map_type;
+  using LO             = typename SparseMatrixType::local_ordinal_type;
+  using KCRS           = typename SparseMatrixType::local_matrix_type;
+  using lno_view_t     = typename KCRS::StaticCrsGraphType::row_map_type::const_type;
+  using lno_nnz_view_t = typename KCRS::StaticCrsGraphType::entries_type::const_type;
+  using scalar_view_t  = typename KCRS::values_type::const_type;
+  using local_mv_type  = typename MultiVectorType::dual_view_type::t_dev;
+  using range_type     = Kokkos::RangePolicy<typename SparseMatrixType::node_type::execution_space, size_t>;
+
+  // Sanity checking: Map Compatibility (A's rowmap matches diagonal's map)
+  TEUCHOS_TEST_FOR_EXCEPTION(!A.getRowMap()->isSameAs(*diagonal.getMap()),
+    std::runtime_error, "Tpetra::Details::extractBlockDiagonal was given incompatible maps");
+
+  int numrows   = diagonal.getLocalLength();
+  int blocksize = diagonal.getNumVectors();
+  diagonal.putScalar(Teuchos::ScalarTraits<typename MultiVectorType::scalar_type>::zero());
+
+  // Get Kokkos versions of objects
+  local_map_type rowmap  = A.getRowMap()->getLocalMap();
+  local_map_type colmap  = A.getRowMap()->getLocalMap();
+  local_mv_type diag     = diagonal.getLocalViewDevice();
+  const KCRS & Amat      = A.getLocalMatrix();
+  lno_view_t Arowptr     = Amat.graph.row_map;
+  lno_nnz_view_t Acolind = Amat.graph.entries;
+  scalar_view_t Avals    = Amat.values;
+
+  Kokkos::parallel_for("Tpetra::extractBlockDiagonal",range_type(0,numrows),[=](const size_t i){
+      LO diag_col   = colmap.getLocalElement(rowmap.getGlobalElement(i));
+      LO blockStart = diag_col - blocksize + 1;
+      LO blockStop  = diag_col + blocksize - 1;
+      
+      for (size_t k = Arowptr(i); k < Arowptr(i+1); k++) {
+        LO col = Acolind(k);
+        if (blockStart <= col && col <= blockStop) {
+          diag(i,col-blockStart) = Avals(k);
```

Comment:
I'm a little bit confused about the purpose of this kernel.  It looks like you're flattening out the blocks into a row of the MultiVector; is that right?

Reply:
Yes